cmd/pebble/cmd_enter.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/jessevdk/go-flags"
+)
+
+const shortEnterHelp = "Run pebble command in new pebble environment"
+const longEnterHelp = `
+The enter command behaves like the run command with the possibility of running
+a single pebble command after starting the service manager. It's intended to be
+used as entrypoint in container images.
+
+When no subcommand is given, the behavior is identical to the run command.
+
+When a subcommand is given, the service manager is started in the background
+without starting "autostart" services (i.e. run --hold, unless --run flag is
+given), and the subcommand is run as a pebble client command that can
+communicate with the service manager instance in the background. The service
+manager is shut down after the commands finishes unless noted otherwise below.
+
+Only a subset of pebble commands is supported as a subcommand:
+
+  exec
+  help
+  ls
+  plan
+  services
+  start
+  version
+
+Furthermore, some commands have restrictions and idiosyncrasies:
+
+  exec, ls, plan, services, version
+    Service manager logging is disabled unless -v/--verbose flag is provided.
+
+  help, ls, plan, services, version
+    Service autostart (the --run flag) is prohobitied.
+
+  start, stop
+    Service manager continues running after the command finishes.
+
+  stop
+    Service autostart (the --run flag) is required.
+`
+
+type enterFlags int
+
+const (
+	enterFlagSilenceLogging enterFlags = 1 << iota
+	enterFlagNoServiceManager
+	enterFlagKeepServiceManager
+	enterFlagRequireServiceAutostart
+	enterFlagProhibitServiceAutostart
+)
+
+type cmdEnter struct {
+	clientMixin
+
+	sharedRunEnterOpts
+	Run        bool `long:"run"`
+	Positional struct {
+		Cmd []string `positional-arg-name:"<subcommand>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	optsHelp := map[string]string{
+		"run": "Start \"autostart\" services before executing subcommand",
+	}
+	for k, v := range sharedRunEnterOptsHelp {
+		optsHelp[k] = v
+	}
+	cmdInfo := addCommand("enter", shortEnterHelp, longEnterHelp, func() flags.Commander { return &cmdEnter{} }, optsHelp, nil)
+	cmdInfo.extra = func(cmd *flags.Command) {
+		cmd.PassAfterNonOption = true
+	}
+}
+
+func commandEnterFlags(commander flags.Commander) (enterFlags enterFlags, supported bool) {
+	supported = true
+	switch commander.(type) {
+	case *cmdExec:
+		enterFlags = enterFlagSilenceLogging
+	case *cmdHelp:
+		enterFlags = enterFlagNoServiceManager
+	case *cmdLs, *cmdPlan, *cmdServices, *cmdVersion:
+		enterFlags = enterFlagSilenceLogging | enterFlagProhibitServiceAutostart
+	case *cmdStart:
+		enterFlags = enterFlagKeepServiceManager
+	case *cmdStop:
+		enterFlags = enterFlagKeepServiceManager | enterFlagRequireServiceAutostart
+	default:
+		supported = false
```

Comment:
I was thinking about allowing each of the commands to have their own flags defined locally, but the clarity of this function is so nice and readable. It indeed makes it feel like a better approach.